Add cholesky(::Diagonal), with tests.

tpapp opened the pull request:
Fixes #27821.

tpapp committed:
Add cholesky(::Diagonal), with tests.

fredrikekre commented:
Should this return a `Cholesky`?

tpapp committed:
Wrap value in Cholesky, fix tests, include order.

tpapp commented:
@fredrikekre : Good catch (I forgot to commit). Also, `Diagonal` needs to be defined, changed order.

The only think I am uncertain about is whether `.U` should be an `UpperTriangular` or to special case it to a `Diagonal`.

tpapp committed:
Move diagonal.jl include near the other types (cosmetic).

andreasnoack commented on stdlib/LinearAlgebra/src/cholesky.jl:
```diff
@@ -174,6 +174,11 @@ function cholesky!(A::StridedMatrix, ::Val{false}=Val(false); check::Bool = true
     end
 end
 
+function cholesky(A::Diagonal, ::Val{false} = Val(false))
```
You'd need to add the `check` keyword here and only throw when it is true and otherwise set the `info` value to the index of the first non-positive element.

tpapp committed:
Handle check=false, add complex w/tests, move include back.

1. Code is moved to diagonal.jl, this looked like the simplest
solution to have all types available for functions.

2. Added case for `check = false`, which does not fail if the diagonal
value is ≤ 0, but records it in `info`.

3. Allow for complex values, only check for `== 0`.

4. Unit tests for `check = false`, also for complex case.

tpapp committed:
Fix deprecated syntax in tests.

tpapp commented:
The test failure looks like a quirk of the CI system. @andreasnoack, is there anything I should improve on to get this merged?

andreasnoack commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -476,3 +476,21 @@ end
+function cholesky(A::Diagonal, ::Val{false} = Val(false); check::Bool = true)
+    info = 0
+    function Dchecked(i, d)
+        if (d isa Complex && d ≠ 0) || (d isa Real && d > 0)
```
In theory, the input might be a `Quaternion` so probably better with
```julia
if isreal(d) && real(d) > 0
```

andreasnoack reviewed (commented):
I think it would be better to define the mutating version `cholesky!`. Then the non-mutating version should be available through the fallbacks.

tpapp committed:
Add cholesky!(::Diagonal), generalize tests for positive definiteness.

1. Control flow is simplified by using a loop.

2. Allow for the possibility of types other than Real or Complex (eg
Quaternion).

3. Add tests `InexactError` for some `cholesky!(::Diagonal{Int})`.

tpapp committed:
Fix missing default.

tpapp commented:
@andreasnoack: thanks for the review, I implemented the changes. The CI failures look unrelated to me.

tpapp commented:
@andreasnoack: can you please take a look at this and consider merging, or tell me if I should do anything else?

StefanKarpinski commented:
Given that this passes tests and looks vaguely reasonable to my unspecialized eye, I will merge by within 24 hours if no review occurs. If someone else beats me to it, please feel free.
